typescript/sdk/src/providers/MultiProvider.ts

```diff
@@ -1,139 +1,272 @@
+        publicRpcUrls[0].http,
+        id,
+      );
     }
-    return chainConnection.signer;
+
+    return this.providers[name] ?? null;
   }
 
   /**
-   * Get signer for a chain
-   * @returns value or null if chain value has not been set
+   * Get an Ethers provider for a given chain name or id
+   * @throws if chain's metadata has not been set
    */
-  tryGetChainSigner(chain: Chain): ethers.Signer | null {
-    return this.tryGet(chain)?.signer ?? null;
+  getProvider(chainNameOrId: ChainName | number): Provider {
+    const provider = this.tryGetProvider(chainNameOrId);
+    if (!provider)
+      throw new Error(`No chain metadata set for ${chainNameOrId}`);
+    return provider;
   }
 
   /**
-   * Get the id for a given chain name
-   * Attempts to use SDK defaults first, otherwise queries network
-   * @throws if chain is invalid or has not been set
+   * Sets an Ethers provider for a given chain name or id
+   * @throws if chain's metadata has not been set
    */
-  getChainId(chain: Chain): number {
-    return this.getChainConnection(chain).id;
+  setProvider(chainNameOrId: ChainName | number, provider: Provider): Provider {
+    const chainName = this.getChainName(chainNameOrId);
+    this.providers[chainName] = provider;
+    return provider;
   }
 
   /**
-   * Create a new MultiProvider which includes the provided chain connection config
+   * Get an Ethers signer for a given chain name or id
    */
-  extendWithChain<New extends Remotes<ChainName, Chain>>(
-    chain: New,
-    chainConnectionConfig: IChainConnection,
-  ): MultiProvider<New & Chain> {
-    const connection = new ChainConnection(chainConnectionConfig);
-    return new MultiProvider<New & Chain>({
-      ...this.chainMap,
-      [chain]: connection,
-    });
+  tryGetSigner(chainNameOrId: ChainName | number): Signer | null {
+    const chainName = this.tryGetChainName(chainNameOrId);
+    if (chainName && this.signers[chainName]) return this.signers[chainName];
+    return null;
+  }
+
+  /**
+   * Get an Ethers signer for a given chain name or id
+   * @throws if chain's metadata or signer has not been set
+   */
+  getSigner(chainNameOrId: ChainName | number): Signer {
+    const chainName = this.getChainName(chainNameOrId);
+    if (this.signers[chainName]) return this.signers[chainName];
+    else throw new Error(`No chain signer set for ${chainNameOrId}`);
+  }
+
+  /**
+   * Get an Ethers signer for a given chain name or id
+   * @throws if chain's metadata or signer has not been set
+   */
+  async getSignerAddress(
+    chainNameOrId: ChainName | number,
+  ): Promise<types.Address> {
+    const signer = this.getSigner(chainNameOrId);
+    const address = await signer.getAddress();
+    return address;
+  }
+
+  /**
+   * Sets an Ethers Signer for a given chain name or id
+   * @throws if chain's metadata has not been set
+   */
+  setSigner(chainNameOrId: ChainName | number, signer: Signer): Signer {
```

Comment:
The old `setSigner` method would use the existing providers as the "base" and we should decide whether to keep that or not?

Reply:
You mean to connect them to providers?